From: Enrico Jorns <ejo@pengutronix.de>
Date: Mon, 26 Jan 2015 14:46:29 +0100
Subject: [PATCH] v4l2-utils: Intersect probed caps with template caps

Template caps might be more strict than probed caps.
For an encoder/decoder element that supports more formats than specified
in its template, these additional formats should be hidden.
---
 sys/v4l2/v4l2-utils.c | 29 +++++++++++++++++++++--------
 1 file changed, 21 insertions(+), 8 deletions(-)

diff --git a/sys/v4l2/v4l2-utils.c b/sys/v4l2/v4l2-utils.c
index c0c509ed1f58..c26a946a1ef5 100644
--- a/sys/v4l2/v4l2-utils.c
+++ b/sys/v4l2/v4l2-utils.c
@@ -241,16 +241,25 @@ gst_v4l2_encoder_subclass_init (gpointer g_class, gpointer data)
   GObjectClass *gobject_class =
       (GObjectClass *) GST_V4L2_VIDEO_ENC_CLASS (g_class);
   GstV4l2VideoCData *cdata = data;
+  GstCaps *tcaps;
 
   GST_V4L2_VIDEO_ENC_CLASS (g_class)->default_device = cdata->device;
 
   /* Note: gst_pad_template_new() take the floating ref from the caps */
+  tcaps = gst_pad_template_get_caps (gst_element_class_get_pad_template
+      (element_class, "sink"));
   gst_element_class_add_pad_template (element_class,
-      gst_pad_template_new ("sink",
-          GST_PAD_SINK, GST_PAD_ALWAYS, cdata->sink_caps));
+      gst_pad_template_new ("sink", GST_PAD_SINK, GST_PAD_ALWAYS,
+          gst_caps_intersect_full (cdata->sink_caps, tcaps,
+              GST_CAPS_INTERSECT_FIRST)));
+  gst_caps_unref (tcaps);
+  tcaps = gst_pad_template_get_caps (gst_element_class_get_pad_template
+      (element_class, "src"));
   gst_element_class_add_pad_template (element_class,
-      gst_pad_template_new ("src",
-          GST_PAD_SRC, GST_PAD_ALWAYS, cdata->src_caps));
+      gst_pad_template_new ("src", GST_PAD_SRC, GST_PAD_ALWAYS,
+          gst_caps_intersect_full (cdata->src_caps, tcaps,
+              GST_CAPS_INTERSECT_FIRST)));
+  gst_caps_unref (tcaps);
 
   gobject_class->get_property =
       GST_DEBUG_FUNCPTR (gst_v4l2_subclass_get_property);
@@ -288,11 +297,15 @@ gst_v4l2_decoder_subclass_init (gpointer g_class, gpointer data)
 
   /* Note: gst_pad_template_new() take the floating ref from the caps */
   gst_element_class_add_pad_template (element_class,
-      gst_pad_template_new ("sink",
-          GST_PAD_SINK, GST_PAD_ALWAYS, cdata->sink_caps));
+      gst_pad_template_new ("sink", GST_PAD_SINK, GST_PAD_ALWAYS,
+          gst_caps_intersect_full (cdata->sink_caps,
+              gst_pad_template_get_caps (gst_element_class_get_pad_template
+                  (element_class, "sink")), GST_CAPS_INTERSECT_FIRST)));
   gst_element_class_add_pad_template (element_class,
-      gst_pad_template_new ("src",
-          GST_PAD_SRC, GST_PAD_ALWAYS, cdata->src_caps));
+      gst_pad_template_new ("src", GST_PAD_SRC, GST_PAD_ALWAYS,
+          gst_caps_intersect_full (cdata->src_caps,
+              gst_pad_template_get_caps (gst_element_class_get_pad_template
+                  (element_class, "src")), GST_CAPS_INTERSECT_FIRST)));
 
   gobject_class->get_property =
       GST_DEBUG_FUNCPTR (gst_v4l2_subclass_get_property);
